utils/result_try.hh

```diff
@@ -0,0 +1,442 @@
+    }
+
+    template<typename F, typename... Args>
+    static auto invoke(F&& f, Args&&... args) {
+        return f(std::forward<Args>(args)...);
+    }
+};
+
+struct futurizing_converter {
+    template<typename T>
+    using wrapped_type = typename seastar::futurize<T>::type;
+
+    template<typename T>
+    static auto wrap(T&& t) {
+        if constexpr (seastar::is_future<T>::value) {
+            return t;
+        } else {
+            return seastar::make_ready_future<T>(std::forward<T>(t));
+        }
+    }
+
+    template<typename F, typename... Args>
+    static auto invoke(F&& f, Args&&... args) {
+        return seastar::futurize_invoke(std::forward<F>(f), std::forward<Args>(args)...);
+    }
+};
+
+// TODO: Should we make a virtual base class for this?
+// This would allow using non-generic lambdas in result_catch and result_catch_dots.
+template<typename Handle, typename R>
+concept ExceptionHandle = ExceptionContainerResult<R> && requires (Handle handle) {
+    { handle.into_result() } -> std::same_as<R>;
+    { handle.into_future() } -> std::same_as<seastar::future<R>>;
+};
+
+template<ExceptionContainerResult R>
+struct failed_result_handle {
+private:
+    R&& _failed_result;
+
+public:
+    failed_result_handle(R&& r) : _failed_result(std::move(r)) {}
+
+    // Returns the failed result. Can be used only once as this moves out
+    // the failed result. This should be called once during the handler's
+    // lifetime, preferably as `return handler.into_failure()`.
+    R into_result() {
+        return std::move(_failed_result);
+    }
+
+    seastar::future<R> into_future() {
+        return seastar::make_ready_future<R>(std::move(_failed_result));
+    }
+};
+
+static_assert(ExceptionHandle<failed_result_handle<dummy_result<>>, dummy_result<>>);
+
+template<ExceptionContainerResult R>
+struct exception_ptr_handle {
+private:
+    std::exception_ptr _eptr;
+
+public:
+    exception_ptr_handle(std::exception_ptr&& eptr) : _eptr(std::move(eptr)) {}
+
+    // Throws the exception.
+    [[noreturn]] R into_result() {
+        std::rethrow_exception(std::move(_eptr));
+    }
+
+    seastar::future<R> into_future() {
+        return seastar::make_exception_future<R>(std::move(_eptr));
+    }
+};
+
+// Check that it is a valid handle with some arbitrary ExceptionContainerResult
+static_assert(ExceptionHandle<exception_ptr_handle<dummy_result<>>, dummy_result<>>);
+
+template<typename Ex, typename Cb>
+struct result_catcher {
+private:
+    Cb _cb;
+
+public:
+    result_catcher(Cb&& cb) : _cb(std::move(cb)) {}
+
+    template<ExceptionContainerResult R, typename Converter, typename Continuation>
+    auto handle_exception_from_result(const auto& ex, R& original_result, Continuation&& cont)
+            -> typename Converter::template wrapped_type<R> {
+        if constexpr (std::is_base_of_v<Ex, std::remove_cvref_t<decltype(ex)>>) {
+            if constexpr (std::is_invocable_v<Cb, const Ex&, failed_result_handle<R>>) {
```

Comment:
Not sure. I think you could pass a `const Ex&` to a function `void foo(Ex ex)`. However it would require a copy constructor to be invoked and I'm not sure how it would work if you have a class derived from `std::exception`.